### PR TITLE
AP_Filesystem: add support for listing @SYS files

### PR DESCRIPTION
**Pull request by bugobliterator**

### CODE REVIEW

File: libraries/AP_Filesystem/AP_Filesystem_Sys.cpp [hunk 1]
```diff
@@ -145,10 +176,61 @@ int32_t AP_Filesystem_Sys::lseek(int fd, int32_t offset, int seek_from)
...
-    memset(stbuf, 0, sizeof(*stbuf));
-    // give fixed size, EOF on read gives real size
-    stbuf->st_size = 1024*1024;
+    DirReadTracker* dtracker = ((DirReadTracker*)dirp);
+    if (dtracker->file_offset > ARRAY_SIZE(sysfs_file_list)) {
+        // we have reached end of list
+        return nullptr;
+    }
+    dtracker->curr_file.d_type = DT_REG;
+    strcpy(dtracker->curr_file.d_name, sysfs_file_list[dtracker->file_offset].name);
```
Comment (tridge) [inline libraries/AP_Filesystem/AP_Filesystem_Sys.cpp:203]: strncpy please

File: libraries/AP_Filesystem/AP_Filesystem_Sys.cpp [hunk 2]
```diff
@@ -145,10 +176,61 @@ int32_t AP_Filesystem_Sys::lseek(int fd, int32_t offset, int seek_from)
...
+    if (pathname == nullptr || stbuf == nullptr || (strlen(pathname) == 0)) {
+        errno = EINVAL;
+        return -1;
+    }
+
+    if (strlen(pathname) == 1 && pathname[0] == '/') {
+        stbuf->st_size = 0; // just a placeholder value
+        return 0;
+    }
+    int pos = file_in_sysfs(&pathname[1]);
```
Comment (tridge) [inline libraries/AP_Filesystem/AP_Filesystem_Sys.cpp:229]: stdint types

File: libraries/AP_Filesystem/AP_Filesystem_Sys.cpp [hunk 3]
```diff
@@ -145,10 +176,61 @@ int32_t AP_Filesystem_Sys::lseek(int fd, int32_t offset, int seek_from)
...
+    if (pos < 0) {
+        errno = ENOENT;
+        return -1;
+    }
```
Comment (tridge) [inline libraries/AP_Filesystem/AP_Filesystem_Sys.cpp:233]: should memset 0 for other fields

File: libraries/AP_Filesystem/AP_Filesystem_Sys.cpp [hunk 4]
```diff
@@ -25,6 +25,30 @@
...
+    {"threads.txt", 1024},
+    {"tasks.txt", 6500},
+#if HAL_MAX_CAN_PROTOCOL_DRIVERS
+    {"can_log.txt", 1024},
+    {"can0_stats.txt", 1024},
+    {"can1_stats.txt", 1024},
+#endif
+};
+
+int AP_Filesystem_Sys::file_in_sysfs(const char *fname) {
```
Comment (tridge) [inline libraries/AP_Filesystem/AP_Filesystem_Sys.cpp:43]: comments please

File: libraries/AP_Filesystem/AP_Filesystem_Sys.cpp [hunk 5]
```diff
@@ -145,10 +176,61 @@ int32_t AP_Filesystem_Sys::lseek(int fd, int32_t offset, int seek_from)
...
+        // no sub directories
+        errno = ENOENT;
+        return nullptr;
+    }
+    DirReadTracker *dtracker = new DirReadTracker;
+    if (dtracker == nullptr) {
+        errno = ENOMEM;
+        return nullptr;
+    }
+    dtracker->file_offset = 0;
```
Comment (tridge) [inline libraries/AP_Filesystem/AP_Filesystem_Sys.cpp:191]: don't need zero

**Review by tridge — CHANGES_REQUESTED** [review]

**Comment by davidbuzz** [comment]
looks like a good change to me, not sure if this impacts ram usage etc tho.

**Review by tridge — APPROVED** [review]

**Review by tridge — APPROVED** [review]

**Pull request merged**